File: ios/mediquo_flutter/Sources/mediquo_flutter/MediquoFlutterPlugin.swift
import Flutter
import MediQuoSDK
import UIKit

/// Flutter plugin bridging the native MediQuo iOS SDK.
///
/// Implements the Pigeon-generated `MediquoHostApi` protocol and forwards every
/// call to a `MediQuo` instance, completing each result with `.success` or a
/// `PigeonError` whose code matches `MediquoErrorCode` on the Dart side.
public class MediquoFlutterPlugin: NSObject, FlutterPlugin, MediquoHostApi {

  private enum ErrorCode {
    static let initializationFailed = "initialization_failed"
    static let authenticationFailed = "authentication_failed"
    static let openFailed = "open_failed"
    static let deauthenticationFailed = "deauthentication_failed"
    static let pushRegistrationFailed = "push_registration_failed"
    static let notInitialized = "not_initialized"
  }

  private var apiKey: String?
  private var mediquo: MediQuo?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let plugin = MediquoFlutterPlugin()
    MediquoHostApiSetup.setUp(
      binaryMessenger: registrar.messenger(),
      api: plugin
    )
  }

  func initialize(
    apiKey: String,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    self.apiKey = apiKey
    completion(.success(()))
  }

  func authenticate(
    clientCode: String,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let apiKey else {
      completion(
        .failure(
          PigeonError(
            code: ErrorCode.notInitialized,
            message: "Initialise the SDK before authenticating.",
            details: nil
          )
        )
      )
      return
    }
    Task {
      do {
        mediquo = try await MediQuo(apiKey: apiKey, userID: clientCode)
        completion(.success(()))
      } catch {
        completion(
          .failure(
            PigeonError(
              code: ErrorCode.authenticationFailed,
              message: error.localizedDescription,
              details: nil
            )
          )
        )
      }
    }
  }

  func openProfessionalList(
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let mediquo else {
      completion(.failure(notInitialised()))
      return
    }
    DispatchQueue.main.async {
      guard let presenter = Self.topViewController() else {
        completion(
          .failure(
            PigeonError(
              code: ErrorCode.openFailed,
              message: "No view controller available to present from.",
              details: nil
            )
          )
        )
        return
      }
      let viewController = mediquo.sdkViewController(for: .professionalList)
      presenter.present(viewController, animated: true)
      completion(.success(()))
    }
  }

  func deauthenticate(
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let mediquo else {
      completion(.failure(notInitialised()))
      return
    }
    Task {
      do {
        try await mediquo.deauthenticateSDK()
        self.mediquo = nil
        completion(.success(()))
      } catch {
        completion(
          .failure(
            PigeonError(
              code: ErrorCode.deauthenticationFailed,
              message: error.localizedDescription,
              details: nil
            )
          )
        )
      }
    }
  }

  func registerPushToken(
    token: String,
    type: PushTokenType,
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let mediquo else {
      completion(.failure(notInitialised()))
      return
    }
    Task {
      do {
        switch type {
        case .fcm:
          try await mediquo.setPushNotificationToken(type: .firebase(token))
        case .apns:
          guard let data = Self.data(fromHex: token) else {
            completion(
              .failure(
                PigeonError(
                  code: ErrorCode.pushRegistrationFailed,
                  message: "The APNs token is not valid hexadecimal.",
                  details: nil
                )
              )
            )
            return
          }
          try await mediquo.setPushNotificationToken(type: .appleAPNS(data))
        }
        completion(.success(()))
      } catch {
        completion(
          .failure(
            PigeonError(
              code: ErrorCode.pushRegistrationFailed,
              message: error.localizedDescription,
              details: nil
            )
          )
        )
      }
    }
  }

  func openFromRemoteNotification(
    payload: [String: Any?],
    completion: @escaping (Result<Void, Error>) -> Void
  ) {
    guard let mediquo else {
      completion(.failure(notInitialised()))
      return
    }
    DispatchQueue.main.async {
      guard let presenter = Self.topViewController() else {
        completion(
          .failure(
            PigeonError(
              code: ErrorCode.openFailed,
              message: "No view controller available to present from.",
              details: nil
            )
          )
        )
        return
      }
      let userInfo = payload.compactMapValues { $0 }
      let viewController = mediquo.getSDKViewController(forRemotePush: userInfo)
      presenter.present(viewController, animated: true)
      completion(.success(()))
    }
  }

  private func notInitialised() -> PigeonError {
    PigeonError(
      code: ErrorCode.notInitialized,
      message: "Authenticate a patient before this operation.",
      details: nil
    )
  }

  private static func topViewController() -> UIViewController? {
    let scene = UIApplication.shared.connectedScenes
      .compactMap { $0 as? UIWindowScene }
      .first { $0.activationState == .foregroundActive }
    let window = scene?.windows.first { $0.isKeyWindow }
    var top = window?.rootViewController
    while let presented = top?.presentedViewController {
      top = presented
    }
    return top
  }

  private static func data(fromHex hex: String) -> Data? {
    guard hex.count.isMultiple(of: 2), !hex.isEmpty else { return nil }
    var data = Data(capacity: hex.count / 2)
    var index = hex.startIndex
    while index < hex.endIndex {
      let next = hex.index(index, offsetBy: 2)
      guard let byte = UInt8(hex[index..<next], radix: 16) else { return nil }
      data.append(byte)
      index = next
    }
    return data
  }
}

File: android/src/main/kotlin/com/mediquo/mediquo_flutter/MediquoFlutterPlugin.kt
package com.mediquo.mediquo_flutter

import android.app.Activity
import android.content.Context
import com.mediquo.chat.MediquoAuthenticateListener
import com.mediquo.chat.MediquoDeAuthenticateListener
import com.mediquo.chat.MediquoInitListener
import com.mediquo.chat.MediquoSDK
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

/**
 * Flutter plugin bridging the native MediQuo Android SDK.
 *
 * Implements the Pigeon-generated [MediquoHostApi] and forwards every call to
 * [MediquoSDK], completing each callback with [Result.success] on success or a
 * [FlutterError] whose code matches `MediquoErrorCode` on the Dart side.
 */
class MediquoFlutterPlugin : FlutterPlugin, ActivityAware, MediquoHostApi {

    private var applicationContext: Context? = null
    private var activity: Activity? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        MediquoHostApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = null
        MediquoHostApi.setUp(binding.binaryMessenger, null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(
        binding: ActivityPluginBinding,
    ) {
        activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun initialize(apiKey: String, callback: (Result<Unit>) -> Unit) {
        val context = applicationContext
        if (context == null) {
            callback(failure(NOT_INITIALIZED, "The plugin is not attached."))
            return
        }
        MediquoSDK.initialize(
            context,
            apiKey,
            object : MediquoInitListener {
                override fun onSuccess() = callback(Result.success(Unit))

                override fun onFailure(message: String?) =
                    callback(failure(INITIALIZATION_FAILED, message))
            },
        )
    }

    override fun authenticate(
        clientCode: String,
        callback: (Result<Unit>) -> Unit,
    ) {
        MediquoSDK.authenticate(
            clientCode,
            object : MediquoAuthenticateListener {
                override fun onSuccess() = callback(Result.success(Unit))

                override fun onFailure(message: String?) =
                    callback(failure(AUTHENTICATION_FAILED, message))
            },
        )
    }

    override fun openProfessionalList(callback: (Result<Unit>) -> Unit) {
        val instance = MediquoSDK.getInstance()
        val currentActivity = activity
        if (instance == null) {
            callback(failure(NOT_INITIALIZED, "Authenticate a patient first."))
            return
        }
        if (currentActivity == null) {
            callback(failure(OPEN_FAILED, "No foreground activity available."))
            return
        }
        runCatching { instance.openProfessionalListActivity(currentActivity) }
            .fold(
                { callback(Result.success(Unit)) },
                { callback(failure(OPEN_FAILED, it.message)) },
            )
    }

    override fun deauthenticate(callback: (Result<Unit>) -> Unit) {
        MediquoSDK.deAuthenticate(
            object : MediquoDeAuthenticateListener {
                override fun onSuccess() = callback(Result.success(Unit))

                override fun onFailure(message: String?) =
                    callback(failure(DEAUTHENTICATION_FAILED, message))
            },
        )
    }

    override fun registerPushToken(
        token: String,
        type: PushTokenType,
        callback: (Result<Unit>) -> Unit,
    ) {
        val instance = MediquoSDK.getInstance()
        if (instance == null) {
            callback(failure(NOT_INITIALIZED, "Authenticate a patient first."))
            return
        }
        runCatching { instance.registerPushToken(token) }
            .fold(
                { callback(Result.success(Unit)) },
                { callback(failure(PUSH_REGISTRATION_FAILED, it.message)) },
            )
    }

    override fun openFromRemoteNotification(
        payload: Map<String, Any?>,
        callback: (Result<Unit>) -> Unit,
    ) {
        val instance = MediquoSDK.getInstance()
        val currentActivity = activity
        if (instance == null) {
            callback(failure(NOT_INITIALIZED, "Authenticate a patient first."))
            return
        }
        if (currentActivity == null) {
            callback(failure(OPEN_FAILED, "No foreground activity available."))
            return
        }
        runCatching { instance.openProfessionalListActivity(currentActivity) }
            .fold(
                { callback(Result.success(Unit)) },
                { callback(failure(OPEN_FAILED, it.message)) },
            )
    }

    private fun failure(code: String, message: String?): Result<Unit> =
        Result.failure(FlutterError(code, message, null))

    private companion object {
        const val INITIALIZATION_FAILED = "initialization_failed"
        const val AUTHENTICATION_FAILED = "authentication_failed"
        const val OPEN_FAILED = "open_failed"
        const val DEAUTHENTICATION_FAILED = "deauthentication_failed"
        const val PUSH_REGISTRATION_FAILED = "push_registration_failed"
        const val NOT_INITIALIZED = "not_initialized"
    }
}
